File: hkrd/api/live.py
"""Conditional replies for the two pages that follow a live market.

A market captured every minute is only worth capturing if something reads it
every minute. Race Day and Model Analysis are snapshots from page load: a card
opened ten minutes before the off shows a price from ten minutes before the
off, which is precisely when it stops being true.

So both pages poll. This is what stops that costing anything.

**How stale the screen can be is bounded by the capture, not by a second
schedule.** `query/market.poll_seconds` returns half the capture interval for
that race, read off the same ladder `jobs/scrape_odds` captures on. Overnight
that is thirty minutes; in the last ten minutes before the off it is fifteen
seconds. The server tells the browser when to come back, on every reply
including a 304, so there is one schedule and it lives with the data.

**A poll that finds nothing new costs two indexed `max()` lookups.** The Race
Day card is 36 KB and ~220ms to assemble; the poll that discovers it has not
changed answers in under a millisecond and sends no body. That is the whole
design: the expensive path runs only when the data actually moved.

ETag rather than a timestamp comparison the client does, because the client
must not be the one deciding what counts as new — and because a browser
already knows how to hold and return an ETag.
"""
from __future__ import annotations

from collections.abc import Callable
from typing import Any

from fastapi import Request
from fastapi.responses import JSONResponse, Response

from hkrd.query import market as market_q

__all__ = ["conditional", "POLL_HEADER"]

# Named rather than inline because the browser reads it on a 304, where there
# is no body to put it in.
POLL_HEADER = "X-Poll-After"


def conditional(request: Request, *, date: str, race_no: int | None,
                build: Callable[[], dict[str, Any]],
                variant: object = None) -> Response:
    """Answer a poll: 304 if nothing changed, otherwise `build()`.

    `build` is not called at all on the unchanged path — that is the point of
    passing it as a callable rather than a result.

    `variant` is anything OTHER than the data that changes the answer — the
    blend's weight is the only one today. Without it, moving the weight slider
    and asking with the previous tag would be answered "nothing changed" and
    the reader would see the old weight's numbers under the new label. The URL
    differs, so a spec-following cache would not confuse them; a client
    carrying its own tag across the change would, and being right only when
    the client is careful is not being right.

    A poll interval rides on both answers. It is advisory: a client that
    ignores it still gets correct data, and a client that honours it stops
    asking hourly questions every thirty seconds.
    """
    current, captured_at, after = market_q.poll_state(date, race_no)
    if variant is not None:
        current = f"{current}|{variant}"
    headers = {POLL_HEADER: str(after),
               # Two different caches must not hold this. The browser's own
               # would serve a stale price with no way to notice, and a shared
               # one would serve one viewer's meeting to another.
               "Cache-Control": "no-cache, private"}
    etag = f'W/"{current}"'
    # A weak validator, and honestly so: it says "this page would render the
    # same", not "these bytes are identical". Nothing here is byte-stable —
    # dict ordering is not a promise — and a strong tag would claim it is.
    if request.headers.get("if-none-match") == etag:
        return Response(status_code=304, headers={**headers, "ETag": etag})

    payload = build()
    payload["poll_after"] = after
    # Carried in the body as well, so the page can say when the price on
    # screen was true rather than when it was fetched. `setdefault`, because a
    # payload that already names its own capture time is the more specific
    # answer -- the card's is the race's, this one is the meeting's.
    payload.setdefault("captured_at", captured_at)
    return JSONResponse(payload, headers={**headers, "ETag": etag})

File: web/assets/live.js
/* live.js — keep a page in step with a market that moves under it.
 *
 * The odds are captured every minute through the last ten minutes before a
 * race. A page that reads them once at load is showing, twenty minutes later,
 * the price from twenty minutes ago — which is exactly the window brief 01
 * says Race Day exists for. So the two pages that read a live price poll.
 *
 * Three things stop that being expensive, and none of them live here:
 *
 *   The server decides the interval. Every reply carries `X-Poll-After`,
 *   computed from the same cadence ladder the capture runs on — half the
 *   capture interval for that race. Overnight that is thirty minutes; inside
 *   the last ten minutes it is fifteen seconds. One schedule, and it lives
 *   with the data rather than being guessed at twice.
 *
 *   A poll that finds nothing new is a 304 with no body, answered from two
 *   indexed lookups. The 36 KB card is assembled only when it changed.
 *
 *   A hidden tab does not poll at all. A card left open on another desktop
 *   overnight would otherwise ask a few hundred pointless questions, and the
 *   first thing anyone does on coming back is look — so it polls immediately
 *   on becoming visible, which is both cheaper and fresher.
 *
 * On failure it backs off rather than hammering: a server that is down does
 * not get a request a second from every open tab, and the page keeps the last
 * price it had rather than blanking. What it must never do is fail silently,
 * so the caller is handed the error to put on screen.
 */

const MIN_SECONDS = 15;
const MAX_BACKOFF_SECONDS = 300;

/** One watched resource. Call `.watch(key, path)` to point it somewhere new. */
export class Live {
  /**
   * @param {(body:object)=>void} onFresh   a changed payload arrived
   * @param {(message:string)=>void} [onError]  a poll failed
   */
  constructor(onFresh, onError) {
    this.onFresh = onFresh;
    this.onError = onError ?? (() => {});
    this.timer = null;
    this.key = null;
    this.path = null;
    this.etag = null;
    this.failures = 0;
    document.addEventListener('visibilitychange', () => {
      if (document.hidden) this.#cancel();
      else if (this.path) this.#schedule(0);
    });
  }

  /** Point at a new resource. `key` identifies it; a repeat is a no-op, so a
   *  re-render that re-declares the same watch does not restart the clock. */
  watch(key, path) {
    if (key === this.key) return;
    this.#cancel();
    this.key = key;
    this.path = path;
    // The ETag belongs to the OLD resource. Carrying it over would ask
    // "changed since?" about a different race and take a 304 for an answer.
    this.etag = null;
    this.failures = 0;
  }

  /** Record the tag and interval from a payload the PAGE fetched itself, so
   *  the first poll after a load is a conditional one rather than a second
   *  full download of what is already on screen. */
  seed(etag, pollAfter) {
    this.etag = etag ?? null;
    this.#schedule(pollAfter ?? 60);
  }

  stop() { this.#cancel(); this.path = null; this.key = null; }

  #cancel() {
    if (this.timer !== null) clearTimeout(this.timer);
    this.timer = null;
  }

  #schedule(seconds) {
    this.#cancel();
    if (!this.path || document.hidden) return;
    this.timer = setTimeout(() => this.#poll(), Math.max(0, seconds) * 1000);
  }

  async #poll() {
    const path = this.path;
    try {
      const res = await fetch(`/api${path}`, {
        // Our own conditional request, not the browser's. With the browser
        // cache in play a 304 can be turned back into a 200 from cache before
        // JS ever sees it, and then "unchanged" and "changed" look identical.
        cache: 'no-store',
        headers: {
          Accept: 'application/json',
          ...(this.etag ? { 'If-None-Match': this.etag } : {}),
        },
      });
      if (path !== this.path) return;          // watched something else mid-flight
      this.failures = 0;
      const after = Number(res.headers.get('X-Poll-After')) || 60;
      const tag = res.headers.get('ETag');
      if (tag) this.etag = tag;
      if (res.status === 304) { this.#schedule(after); return; }
      if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
      const body = await res.json();
      this.onFresh(body);
      this.#schedule(body.poll_after ?? after);
    } catch (e) {
      if (path !== this.path) return;
      this.failures += 1;
      this.onError(e.message);
      // Doubling, capped. A server that is down must not be asked once a
      // second by every tab someone left open.
      this.#schedule(Math.min(MIN_SECONDS * 2 ** this.failures,
                              MAX_BACKOFF_SECONDS));
    }
  }
}
